addons/product_margin/models/product_product.py

```diff
@@ -140,8 +144,10 @@ def _compute_product_margin_fields_values(self, field_names=None):
                 """.format(self.env['res.currency']._select_companies_rates())
         invoice_types = ('out_invoice', 'out_refund')
         self.env.cr.execute(sqlstr, (tuple(self.ids), states, payment_states, invoice_types, date_from, date_to, company_id))
-        for product_id, avg, qty, total, sale in self.env.cr.fetchall():
-            res[product_id]['sale_avg_price'] = avg and avg or 0.0
+        for product_id, avg, qty, total, sale, avg_refund in self.env.cr.fetchall():
+            avg = avg or 0
+            avg_refund = avg_refund or 0
+            res[product_id]['sale_avg_price'] = avg - avg_refund
```

Comment:
For the life of me this maths just does not square with my skull. How can you take the average price of sales and subtract the average price of refunds and get a sensible result unless the in/out quantities exactly match. I mean it isn't called weighted average cost for nothing. I have read maybe 20 times and I just cannot see how this can be right.

In any case I am a bit anti this, because if a quantity is returned in full, then the amount sold is 0. There is no average price of a qty of 0. In the cases where refund amounts differ then it is usually 1. The goods were returned and a restocking fee or similar was charged, in which case represent explicitly as a seperate line or 2. The wrong price was charged, which Odoo does not handle well, but the correct way would be either -1 at Old Price, +1 at new price or a credit and reinvoice. This is how accounts/accountants have done in different systems for decades.

Reply:
For me and my customer it is quite clear. :-)
But i didn't described it well. 

There use case is the following. 
They selling and invoicing a chair for 100 €. 
The customer receives it and sees a little damage. 

They and the customer agree on not returning the chair but simply refund some money. 
In this case a refund is created out of the invoice, which has by default the same qty as the invoice. 
They simply change the amount of the refund for example to 20 €. 

At the end the selled still 1 chair for 80 € so the avg price should be 80 € instead of 0 € which is now the case.

Reply:
Then you are doing what is known as a price credit. Case 2 I listed. As I say all systems for decades have had to handle this and there are really only 2 ways to be exactly accurate. A credit with 1 in, 1 out, or a credit and reinvoice. There are 2 common approaches. Both of which I listed, but simply doing a credit note with a qty of 1 is incorrect, because you are not crediting a qty of 1, you are crediting a quantity of 0. 

So really this is an entry error. You are crediting too much quantity. Your fix papers over 1 little symptom of doing the process wrong. However qty_invoiced, and a whole bunch of reports will also be wrong. It is best just to do price credits properly in the first place.

Reply:
Thanks for your detailed answers. My customer did it always like the way described. From an accounting side this is no problem. Its more that odoo handles it not well as you already said. 
Of course qty_invoiced should be wrong, but on not on this project, i will investigate that and will fix it. 
I checked v17 there is no possibility to create a partial credit_note anymore. There is only full credit note and new invoice possible.